Libraries/Opc.Ua.PubSub/IntervalRunner.cs

```diff
@@ -154,30 +155,53 @@ protected virtual void Dispose(bool disposing)
         /// </summary>
         private async Task ProcessAsync()
         {
+            lock (m_lock)
+            {
+                m_nextPublishTick = HiResClock.Ticks;
+            }
             do
             {
-                int sleepCycle = 0;
-                DateTime now = DateTime.UtcNow;
-                DateTime nextPublishTime = DateTime.MinValue;
-
-                lock (m_lock)
+                if (m_cancellationToken.IsCancellationRequested)
                 {
-                    sleepCycle = Convert.ToInt32(m_interval);
+                    break;
+                }
 
-                    nextPublishTime = m_nextPublishTime;
+                long nowTick = HiResClock.Ticks;
+                double nextPublishTick = 0;
+
+                lock(m_lock)
+                {
+                    nextPublishTick = m_nextPublishTick;
                 }
 
-                if (nextPublishTime > now)
+                double sleepCycle = (nextPublishTick - nowTick) / HiResClock.TicksPerMillisecond;
+                if (sleepCycle > 16)
                 {
-                    sleepCycle = (int)Math.Min((nextPublishTime - now).TotalMilliseconds, sleepCycle);
-                    sleepCycle = (int)Math.Max(kMinInterval, sleepCycle);
+                    // Use Task.Delay if sleep cycle is larger
                     await Task.Delay(TimeSpan.FromMilliseconds(sleepCycle), m_cancellationToken.Token).ConfigureAwait(false);
-                }
 
+                    // Still ticks to consume (spurious wakeup too early), improbable
+                    nowTick = HiResClock.Ticks;
+                    if (nowTick < nextPublishTick)
+                    {
+                        while (HiResClock.Ticks < nextPublishTick)
+                        {
+                            // Busy-wait and avoid overhead of Task.Delay for verry small wait times
```

Comment:
you could add a Thread.SpinWait(1000) or so to keep the CPU busy. the clock/timer may have a lock and polling it may cause some contention with other threads.

Reply:
Wouldn't a SpinWait.SpinUntil(() => HiResClock.Ticks >= nextPublishTick) be better even with actively checking the clock due to it's addaptive nature ?